panic: interface conversion: interface {} is nil, not *commands.DeadlettersCountResponse in getDeadlettersCount under concurrent Prometheus scrape

## Summary

`actor.WithMetrics()` causes a process crash when Prometheus scrapes `/metrics` concurrently with high actor spawn/stop throughput. The panic occurs in `getDeadlettersCount` due to an unguarded type assertion on a potentially nil response.

## Panic

```
panic: interface conversion: interface {} is nil, not *commands.DeadlettersCountResponse

goroutine 26493 [running]:
github.com/tochemey/goakt/v4/actor.(*PID).getDeadlettersCount(0x7c3c4413860, {0x2ffa858, 0x4638bc0})
        github.com/tochemey/goakt/v4@v4.2.5/actor/pid.go:2962
github.com/tochemey/goakt/v4/actor.(*PID).registerMetrics.func1({0x2ffa858, 0x4638bc0}, {0x2ff6bd0, 0x7c3d7efb050})
        github.com/tochemey/goakt/v4@v4.2.5/actor/pid.go:3166
go.opentelemetry.io/otel/sdk/metric.(*meter).RegisterCallback.func1(...)
go.opentelemetry.io/otel/sdk/metric.(*pipeline).produce(...)
go.opentelemetry.io/otel/sdk/metric.(*ManualReader).Collect(...)
go.opentelemetry.io/otel/exporters/prometheus.(*collector).Collect(...)
github.com/prometheus/client_golang/prometheus.(*Registry).Gather.func1()
```

## Root Cause

There are two compounding issues in `pid.go`:

### Issue 1 — TOCTOU race in `getDeadlettersCount`

```go
// pid.go:2956
if to.IsRunning() {
    message, _ := from.Ask(ctx, to, message, DefaultAskTimeout)
    deadlettersCount := message.(*commands.DeadlettersCountResponse) // unsafe
    return deadlettersCount.TotalCount
}
```

`IsRunning()` is not atomic with `Ask()`. If the deadletter actor transitions to stopping between the guard and the `Ask` completing, `Ask` returns `(nil, ErrDead)`. The error is silently discarded and the bare assertion on `nil` panics.

### Issue 2 — OTel Registration handle is discarded

```go
// pid.go:3165
_, err = meter.RegisterCallback(func(...) error {
    ...
    observer.ObserveInt64(metrics.DeadlettersCount(), pid.getDeadlettersCount(ctx), ...)
    ...
}, ...)
```

The `Registration` returned by `RegisterCallback` is dropped (`_`). The OTel SDK keeps this callback alive for the lifetime of the meter — it fires on every Prometheus scrape for every PID that ever existed, including stopped/killed actors. Under high actor churn this accumulates many callbacks all racing against the deadletter actor's availability.

## Note

Interestingly, `getSetDeadlettersCount` on `actorSystem` (line ~3403) already uses the safe pattern:

```go
if deadlettersCount, ok := reply.(*commands.DeadlettersCountResponse); ok && deadlettersCount != nil {
    x.deadlettersCounter.Store(uint64(deadlettersCount.TotalCount))
}
```

The per-PID `getDeadlettersCount` should follow the same pattern.

## Suggested Fix

```go
func (pid *PID) getDeadlettersCount(ctx context.Context) int64 {
    // ...
    if to.IsRunning() {
        resp, err := from.Ask(ctx, to, message, DefaultAskTimeout)
        if err != nil || resp == nil {
            return 0
        }
        deadlettersCount, ok := resp.(*commands.DeadlettersCountResponse)
        if !ok {
            return 0
        }
        return deadlettersCount.TotalCount
    }
    return 0
}
```

The Registration leak (Issue 2) would ideally be fixed separately by storing the handle and calling `Registration.Unregister()` when the PID shuts down.

## Version

GoAkt v4.2.5

--- a/actor/fixtures_test.go
+++ b/actor/fixtures_test.go
@@ -142,6 +142,23 @@ func (sb *safeBuffer) Reset() {
 	sb.buf.Reset()
 }
 
+// wrongDeadletterResponder replies to every request with a message type other
+// than *commands.DeadlettersCountResponse. It is used to drive the type
+// assertion guard in getDeadlettersCount.
+type wrongDeadletterResponder struct{}
+
+var _ Actor = (*wrongDeadletterResponder)(nil)
+
+func (a *wrongDeadletterResponder) PreStart(*Context) error { return nil }
+func (a *wrongDeadletterResponder) PostStop(*Context) error { return nil }
+func (a *wrongDeadletterResponder) Receive(ctx *ReceiveContext) {
+	switch ctx.Message().(type) {
+	case *PostStart:
+	default:
+		ctx.Response(new(testpb.Reply))
+	}
+}
+
 // MockActor is an actor that helps run various test scenarios
 type MockActor struct{}
 
@@ -2384,12 +2401,25 @@ func (m *manualMeterProvider) Meter(_ string, _ ...otelmetric.MeterOption) otelm
 
 type manualMeter struct {
 	otelmetric.Meter
-	callbacks []otelmetric.Callback
+	callbacks    []otelmetric.Callback
+	unregistered int
 }
 
 func (m *manualMeter) RegisterCallback(cb otelmetric.Callback, _ ...otelmetric.Observable) (otelmetric.Registration, error) {
 	m.callbacks = append(m.callbacks, cb)
-	return noopmetric.Registration{}, nil
+	return &trackingRegistration{meter: m}, nil
+}
+
+// trackingRegistration records when a callback registration is unregistered so
+// tests can assert that metrics callbacks are released on actor shutdown.
+type trackingRegistration struct {
+	noopmetric.Registration
+	meter *manualMeter
+}
+
+func (r *trackingRegistration) Unregister() error {
+	r.meter.unregistered++
+	return nil
 }
 
 // immediateMeter triggers the callback at registration time.

--- a/actor/pid.go
+++ b/actor/pid.go
@@ -218,6 +218,11 @@ type PID struct {
 	singletonSpec *singletonSpec
 
 	metricProvider *metric.Provider
+
+	// metricRegistration holds the OTel callback registration so it can be
+	// unregistered when the actor stops. Guarded by fieldsLocker. Without this
+	// the callback outlives the actor and keeps firing on every scrape.
+	metricRegistration otelmetric.Registration
 }
 
 var (
@@ -2562,6 +2567,10 @@ func (pid *PID) unsetBehaviorStacked() {
 func (pid *PID) doStop(ctx context.Context) error {
 	pid.cancelInFlightRequests(gerrors.ErrRequestCanceled)
 
+	// stop the OTel metrics callback so it no longer fires on scrapes once the
+	// actor is gone, preventing callback accumulation under high actor churn.
+	pid.unregisterMetrics()
+
 	defer func() {
 		pid.setState(runningState, false)
 		pid.reset()
@@ -2954,12 +2963,18 @@ func (pid *PID) getDeadlettersCount(ctx context.Context) int64 {
 		}
 	)
 	if to.IsRunning() {
-		// ask the deadletter actor for the count
-		// using the default ask timeout
-		// note: no need to check for error because this call is internal
-		message, _ := from.Ask(ctx, to, message, DefaultAskTimeout)
-		// cast the response received from the deadletter
-		deadlettersCount := message.(*commands.DeadlettersCountResponse)
+		// ask the deadletter actor for the count using the default ask timeout.
+		// IsRunning is not atomic with Ask: if the deadletter actor transitions
+		// to stopping between the guard and the reply, Ask returns (nil, ErrDead).
+		// Discard the count in that case instead of asserting on a nil reply.
+		resp, err := from.Ask(ctx, to, message, DefaultAskTimeout)
+		if err != nil || resp == nil {
+			return 0
+		}
+		deadlettersCount, ok := resp.(*commands.DeadlettersCountResponse)
+		if !ok || deadlettersCount == nil {
+			return 0
+		}
 		return deadlettersCount.TotalCount
 	}
 	return 0
@@ -3156,7 +3171,11 @@ func (pid *PID) registerMetrics() error {
 			otelmetric.WithAttributes(attribute.String("actor.address", pid.ID())),
 		}
 
-		_, err = meter.RegisterCallback(func(ctx context.Context, observer otelmetric.Observer) error {
+		// unregister any previous callback before registering a new one so
+		// that restarting an actor does not leak a second live registration.
+		pid.unregisterMetrics()
+
+		registration, err := meter.RegisterCallback(func(ctx context.Context, observer otelmetric.Observer) error {
 			observer.ObserveInt64(metrics.ChildrenCount(), int64(pid.ChildrenCount()), observeOptions...)
 			observer.ObserveInt64(metrics.StashSize(), int64(pid.StashSize()), observeOptions...)
 			observer.ObserveInt64(metrics.RestartCount(), int64(pid.RestartCount()), observeOptions...)
@@ -3177,10 +3196,31 @@ func (pid *PID) registerMetrics() error {
 			metrics.FailureCount(),
 			metrics.ReinstateCount(),
 		)
+		if err != nil {
+			return err
+		}
 
-		return err
+		pid.fieldsLocker.Lock()
+		pid.metricRegistration = registration
+		pid.fieldsLocker.Unlock()
 	}
 	return nil
+}
+
+// unregisterMetrics unregisters the OTel callback previously installed by
+// registerMetrics, if any. It is safe to call multiple times and when metrics
+// were never registered.
+func (pid *PID) unregisterMetrics() {
+	pid.fieldsLocker.Lock()
+	registration := pid.metricRegistration
+	pid.metricRegistration = nil
+	pid.fieldsLocker.Unlock()
+
+	if registration != nil {
+		// the error returned by Unregister is non-actionable here; the handle
+		// is discarded regardless so the callback stops firing.
+		_ = registration.Unregister()
+	}
 }
 
 // remoteTell sends a message to an actor remotely without expecting any reply

--- a/actor/pid_test.go
+++ b/actor/pid_test.go
@@ -4658,6 +4658,37 @@ func TestNewPID(t *testing.T) {
 		}
 		require.NotEmpty(t, observer.records)
 	})
+	t.Run("With metrics callback unregistered on shutdown", func(t *testing.T) {
+		ctx := context.Background()
+
+		prevProvider := otel.GetMeterProvider()
+		meterProvider := newManualMeterProvider()
+		otel.SetMeterProvider(meterProvider)
+		t.Cleanup(func() { otel.SetMeterProvider(prevProvider) })
+
+		sys, err := NewActorSystem("testSys",
+			WithLogger(log.DiscardLogger),
+			WithMetrics())
+		require.NoError(t, err)
+		require.NotNil(t, sys)
+
+		require.NoError(t, sys.Start(ctx))
+		t.Cleanup(func() { require.NoError(t, sys.Stop(ctx)) })
+
+		manual, ok := meterProvider.meter.(*manualMeter)
+		require.True(t, ok)
+
+		pid, err := sys.Spawn(ctx, "metrics-actor", NewMockActor())
+		require.NoError(t, err)
+		require.NotNil(t, pid)
+
+		before := manual.unregistered
+		require.NoError(t, pid.Shutdown(ctx))
+
+		// stopping the actor must release its metrics callback so it no longer
+		// fires on subsequent scrapes.
+		require.Equal(t, before+1, manual.unregistered)
+	})
 }
 
 func TestLogger(t *testing.T) {
@@ -4716,6 +4747,52 @@ func TestDeadletterCountMetric(t *testing.T) {
 	err = sys.Stop(ctx)
 	assert.NoError(t, err)
 }
+
+func TestDeadletterCountNoPanicWhenAskFails(t *testing.T) {
+	ctx := context.TODO()
+	sys, err := NewActorSystem("testSys", WithLogger(log.DiscardLogger))
+	require.NoError(t, err)
+
+	require.NoError(t, sys.Start(ctx))
+
+	pause.For(time.Second)
+
+	actorRef, err := sys.Spawn(ctx, "actorName", NewMockActor())
+	require.NoError(t, err)
+	require.NotNil(t, actorRef)
+
+	// the deadletter actor is still running, but a cancelled context makes the
+	// internal Ask return (nil, err). The count must be reported as 0 instead
+	// of panicking on a type assertion against a nil reply (see issue #1189).
+	cancelledCtx, cancel := context.WithCancel(ctx)
+	cancel()
+
+	require.NotPanics(t, func() {
+		require.EqualValues(t, 0, actorRef.getDeadlettersCount(cancelledCtx))
+	})
+
+	// an unexpected reply type from the deadletter actor must also report 0
+	// rather than panicking on the type assertion.
+	wrongResponder, err := sys.Spawn(ctx, "wrongResponder", &wrongDeadletterResponder{})
+	require.NoError(t, err)
+
+	system, ok := sys.(*actorSystem)
+	require.True(t, ok)
+	system.locker.Lock()
+	system.deadletter = wrongResponder
+	system.locker.Unlock()
+
+	require.NotPanics(t, func() {
+		require.EqualValues(t, 0, actorRef.getDeadlettersCount(ctx))
+	})
+
+	// once the deadletter actor is no longer running the count is reported as 0.
+	require.NoError(t, wrongResponder.Shutdown(ctx))
+	require.EqualValues(t, 0, actorRef.getDeadlettersCount(ctx))
+
+	require.NoError(t, sys.Stop(ctx))
+}
+
 func TestWatch(t *testing.T) {
 	ctx := context.TODO()
 	host := "127.0.0.1"

